## Image Recognition with Deep Learning with Keras

Image recognition of handwritten digits, using the mnist data set.

In [1]:
import pandas as pd
import numpy as np
import mnist
import keras
from keras.layers import Dense
from keras.models import Sequential
from keras.utils import to_categorical

Using TensorFlow backend.


## Data

The data set is the mnist sample of 60,000 handwritten digits. Each sample is a 28x28 two dimensional array of pixels. Each value in the array is the pixel value and the position in the array represents the position in the actual 2D space where the digit was written.

In our data set, the two dimensional data points are flattened to a one dimensional array of 28x28=784 points. The target set denotes what digit the sample represents, 0-9. Thus, there are ten categories for the target set. Using one-hot-encoding, the target set becomes an array of 10 elements.

The test set is designed, specifically, so that none of the images was drawn by any of the drawers used in the train set.

In [2]:
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)


Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


In [3]:
X_train=mnist.train.images
y_train=mnist.train.labels

X_valid=mnist.validation.images
y_valid=mnist.validation.labels

X_test=mnist.test.images
y_test=mnist.test.labels

In [4]:
#View the shapes of the data sets
print("Shapes")
print("X_train: {}\ny_train: {}\nX_valid: {}\ny_valid: {}\nX_test: {},\ny_test: {}".format(X_train.shape,
                                                                                      y_train.shape,
                                                                                      X_valid.shape,
                                                                                      y_valid.shape,
                                                                                      X_test.shape,
                                                                                      y_test.shape))

Shapes
X_train: (55000, 784)
y_train: (55000, 10)
X_valid: (5000, 784)
y_valid: (5000, 10)
X_test: (10000, 784),
y_test: (10000, 10)


## Keras Model

In [5]:
#Create an instance of Sequential, as our model
model1=Sequential()

In [6]:
#Create two hidden layers with 300 nodes, and an output layer with 10 nodes
model1.add(Dense(300,activation="relu",input_shape=(784,)))
model1.add(Dense(300,activation="relu"))
model1.add(Dense(10,activation="softmax"))

In [7]:
#Create an early stopping callback. This will stop the learning if it is not improving. Set the patience to two.
#The learning will stop if accuracy hasn't improved two epochs in a row
from keras.callbacks import EarlyStopping
early_stopping_monitor=EarlyStopping(patience=2)

In [8]:
#Compile 
model1.compile(optimizer="adam",loss="categorical_crossentropy",metrics=["accuracy"])

In [9]:
#Fit the model
model1.fit(X_train,y_train,epochs=50,callbacks=[early_stopping_monitor])

Epoch 1/50
55000/55000 [==============================] - 15s 272us/step - loss: 0.2049 - acc: 0.9384
Epoch 2/50
  416/55000 [..............................] - ETA: 16s - loss: 0.0852 - acc: 0.9663

C:\Users\Mike\Anaconda3\lib\site-packages\keras\callbacks.py:497: RuntimeWarning: Early stopping conditioned on metric `val_loss` which is not available. Available metrics are: loss,acc
  (self.monitor, ','.join(list(logs.keys()))), RuntimeWarning


55000/55000 [==============================] - 15s 264us/step - loss: 0.0871 - acc: 0.9736
Epoch 3/50
55000/55000 [==============================] - 14s 256us/step - loss: 0.0587 - acc: 0.9807
Epoch 4/50
55000/55000 [==============================] - 14s 257us/step - loss: 0.0434 - acc: 0.9857
Epoch 5/50
55000/55000 [==============================] - 14s 253us/step - loss: 0.0348 - acc: 0.9889
Epoch 6/50
55000/55000 [==============================] - 17s 300us/step - loss: 0.0291 - acc: 0.99061s - loss:
Epoch 7/50
55000/55000 [==============================] - 16s 296us/step - loss: 0.0265 - acc: 0.9914
Epoch 8/50
55000/55000 [==============================] - 17s 307us/step - loss: 0.0204 - acc: 0.9930
Epoch 9/50
55000/55000 [==============================] - 17s 313us/step - loss: 0.0184 - acc: 0.9943
Epoch 10/50
55000/55000 [==============================] - 23s 411us/step - loss: 0.0180 - acc: 0.9942
Epoch 11/50
55000/55000 [==============================] - 22s 397us/step - loss: 

In [10]:
#Evaluate on train and test sets
print("Train Accuracy: {}\nValidation Accuracy: {}".format(model1.evaluate(X_train,y_train),
                                                    model1.evaluate(X_valid,y_valid)))

5000/5000 [==============================] - 0s 63us/step
Train Accuracy: [0.0025840035264986828, 0.99945454545454548]
Validation Accuracy: [0.1864383040974065, 0.98199999999999998]


It looks like we have a very accurate model, comparable to industry standards. Let us check the validation accuracy on a model with the same parameters, but trained with fewer epochs. 50 epochs takes about 20 minutes to train on my laptop (which is not a long time for business applications, but is a long time for the sake of practice.

In [11]:
model2=Sequential()

In [12]:
#Create two hidden layers with 300 nodes, and an output layer with 10 nodes
model2.add(Dense(300,activation="relu",input_shape=(784,)))
model2.add(Dense(300,activation="relu"))
model2.add(Dense(10,activation="softmax"))
#Compile 
model2.compile(optimizer="adam",loss="categorical_crossentropy",metrics=["accuracy"])

In [13]:
#fit the model with 20 epochs
model2.fit(X_train,y_train,epochs=20,callbacks=[early_stopping_monitor],verbose=False)

C:\Users\Mike\Anaconda3\lib\site-packages\keras\callbacks.py:497: RuntimeWarning: Early stopping conditioned on metric `val_loss` which is not available. Available metrics are: loss,acc
  (self.monitor, ','.join(list(logs.keys()))), RuntimeWarning


In [14]:
#Evaluate model2 on train and test sets
print("Train Accuracy: {}\nValidation Accuracy: {}".format(model2.evaluate(X_train,y_train),
                                                    model2.evaluate(X_valid,y_valid)))

5000/5000 [==============================] - 0s 69us/step
Train Accuracy: [0.0091619150468506879, 0.99752727272727271]
Validation Accuracy: [0.12433379970914102, 0.98019999999999996]


We see that we can half our epochs and still achieve similar accuracy. For the sake of model selection and practice, let's use this lower number of epochs for training.

Let's try a couple more models. The first one will be a model with 2 hidden layers of 800 units, and the second will have 3 hidden layers of 300 units.

In [15]:
model3=Sequential()
model3.add(Dense(800,activation="relu",input_shape=(784,)))
model3.add(Dense(800,activation="relu"))
model3.add(Dense(10,activation="softmax"))
#Compile 
model3.compile(optimizer="adam",loss="categorical_crossentropy",metrics=["accuracy"])
#fit the model with 20 epochs
model3.fit(X_train,y_train,epochs=20,callbacks=[early_stopping_monitor],verbose=False)

C:\Users\Mike\Anaconda3\lib\site-packages\keras\callbacks.py:497: RuntimeWarning: Early stopping conditioned on metric `val_loss` which is not available. Available metrics are: loss,acc
  (self.monitor, ','.join(list(logs.keys()))), RuntimeWarning


In [16]:
model4=Sequential()
model4.add(Dense(300,activation="relu",input_shape=(784,)))
model4.add(Dense(300,activation="relu"))
model4.add(Dense(300,activation="relu"))
model4.add(Dense(10,activation="softmax"))
#Compile 
model4.compile(optimizer="adam",loss="categorical_crossentropy",metrics=["accuracy"])
#fit the model with 20 epochs
model4.fit(X_train,y_train,epochs=20,callbacks=[early_stopping_monitor],verbose=False)

C:\Users\Mike\Anaconda3\lib\site-packages\keras\callbacks.py:497: RuntimeWarning: Early stopping conditioned on metric `val_loss` which is not available. Available metrics are: loss,acc
  (self.monitor, ','.join(list(logs.keys()))), RuntimeWarning


In [19]:
#Accuracies of models 3 and 4
print("Model 3:\nTrain Accuracy: {}\nValidation Accuracy: {}".format(model3.evaluate(X_train,y_train,verbose=0),
                                                    model3.evaluate(X_valid,y_valid,verbose=0)))
print("\n\nModel 4:\nTrain Accuracy: {}\nValidation Accuracy: {}".format(model4.evaluate(X_train,y_train,verbose=0),
                                                    model4.evaluate(X_valid,y_valid,verbose=0)))

Model 3:
Train Accuracy: [0.015142077130134542, 0.9965090909090909]
Validation Accuracy: [0.16172480549342036, 0.98160000000000003]


Model 4:
Train Accuracy: [0.0066458705740710606, 0.99812727272727275]
Validation Accuracy: [0.12772725383046699, 0.98019999999999996]


We are not seeing significant improvements with the other models. Assuming all models to be equal, let's choose model 1 to be our model of choice and evaluate it on the test set.

In [21]:
print("Model 1 test accuracy: {}".format(model1.evaluate(X_test,y_test,verbose=0)[1]))

Model 1 test accuracy: 0.9812


## Conclusion

Our model of two hidden layers of 300 nodes each had a test accuracy of 98%. This is similar to the best results achieved for deep learning models, published in the literature.

Increasing layers, nodes, or epochs beyond 20, did not yield substantial increases in accuracy.